sfxclient/httpsink.go

```diff
@@ -243,3 +265,124 @@ func (h *HTTPDatapointSink) encodePostBodyProtobufV2(datapoints []*datapoint.Dat
 	}
 	return body, nil
 }
+
+// HTTPEventSink will accept signalfx events and forward them to SignalFx via HTTP.
+type HTTPEventSink struct {
+	AuthToken      string
+	UserAgent      string
+	Endpoint       string
+	Client         http.Client
+	protoMarshaler func(pb proto.Message) ([]byte, error)
+
+	stats struct {
+		readingBody int64
+	}
+}
+
+// NewHTTPEventSink creates a default NewHTTPDatapointSink using package level constants as
+// defaults, including an empty auth token.  If sending directly to SiganlFx, you will be required
+// to explicitly set the AuthToken
+func NewHTTPEventSink() *HTTPEventSink {
+	return &HTTPEventSink{
+		UserAgent: DefaultUserAgent,
+		Endpoint:  EventIngestEndpointV2,
+		Client: http.Client{
+			Timeout: DefaultTimeout,
+		},
+		protoMarshaler: proto.Marshal,
+	}
+}
+
+// AddEvents forwards the datapoints to SignalFx.
+func (h *HTTPEventSink) AddEvents(ctx context.Context, points []*event.Event) (err error) {
+	if len(points) == 0 {
+		return nil
+	}
+	if ctx.Err() != nil {
+		return errors.Annotate(ctx.Err(), "context already closed")
+	}
+	body, err := h.encodePostBodyProtobufV2(points)
+	if err != nil {
+		return errors.Annotate(err, "cannot encode datapoints into protocol buffers")
+	}
+	req, err := http.NewRequest("POST", h.Endpoint, bytes.NewReader(body))
+	if err != nil {
+		return errors.Annotatef(err, "cannot parse new HTTP request to %s", h.Endpoint)
+	}
+	req.Header.Set("Content-Type", "application/x-protobuf")
+	req.Header.Set(TokenHeaderName, h.AuthToken)
+	req.Header.Set("User-Agent", h.UserAgent)
+	req.Header.Set("Connection", "Keep-Alive")
+
+	return h.withCancel(ctx, req)
+}
+
+func (h *HTTPEventSink) encodePostBodyProtobufV2(events []*event.Event) ([]byte, error) {
+	evs := make([]*com_signalfx_metrics_protobuf.Event, 0, len(events))
+	for _, ev := range events {
+		evs = append(evs, h.coreEventToProtobuf(ev))
+	}
+	msg := &com_signalfx_metrics_protobuf.EventUploadMessage{
+		Events: evs,
+	}
+	body, err := h.protoMarshaler(msg)
+	if err != nil {
+		return nil, errors.Annotate(err, "protobuf marshal failed")
+	}
+	return body, nil
+}
+
+func (h *HTTPEventSink) coreEventToProtobuf(event *event.Event) *com_signalfx_metrics_protobuf.Event {
+	var ts int64
+	if event.Timestamp.IsZero() {
+		ts = 0
+	} else {
+		ts = event.Timestamp.UnixNano() / time.Millisecond.Nanoseconds()
+	}
+	etype := event.EventType
+	ecat := toEC(event.Category)
+	ev := &com_signalfx_metrics_protobuf.Event{
+		EventType:  &etype,
+		Category:   &ecat,
+		Dimensions: mapToDimensions(event.Dimensions),
+		Timestamp:  &ts,
+	}
+	for k, v := range event.Properties {
```

Comment:
feel like this could be mapToProperties since it's copypasta from coreDatapointToProtobuf